tests/test_view_events.py

```diff
@@ -0,0 +1,100 @@
+import asyncio
+
+from lona import LonaView
+
+
+def setup_app(app):
+
+    @app.route('/view-1/')
+    class View1(LonaView):
+        def handle_request(self, request):
+            if 'fire' in request.GET:
+                self.fire_view_event('view1', {'view1': True})
+
+        def on_view_event(self, view_event):
+            self.server.state['view1'].append(
+                (view_event.name, view_event.data),
+            )
+
+    @app.route('/view-2/')
+    class View2(LonaView):
+        def handle_request(self, request):
+            if 'fire' in request.GET:
+                self.fire_view_event('view2', {'view2': True})
+
+        def on_view_event(self, view_event):
+            self.server.state['view2'].append(
+                (view_event.name, view_event.data),
+            )
+
+    @app.route('/broadcast/')
+    class FireBroadcastEvent(LonaView):
+        def handle_request(self, request):
+            self.server.fire_view_event('broadcast', {'broadcast': True})
+
+
+async def test_view_events(lona_app_context):
+    """
+    This test tests view events by creating two views that send events to all
+    of their instances, and one view that sends events to all views in the
+    system.
+    The test is successful if both of the first two views only received their
+    own events, and the broadcast events, and no event from view 1 leaked to
+    view 2 or vice versa.
+    """
+
+    from playwright.async_api import async_playwright
+
+    context = await lona_app_context(setup_app)
+
+    async with async_playwright() as playwright:
+        context.server.state['view1'] = []
+        context.server.state['view2'] = []
+
+        browser = await playwright.chromium.launch()
+        browser_context = await browser.new_context()
+
+        # view 1
+        view1_page1 = await browser_context.new_page()
+        view1_page2 = await browser_context.new_page()
+        view1_page3 = await browser_context.new_page()
+
+        await view1_page1.goto(context.make_url('/view-1/'))
+        await view1_page2.goto(context.make_url('/view-1/'))
+        await view1_page3.goto(context.make_url('/view-1/?fire'))
+
+        # view 2
+        view2_page1 = await browser_context.new_page()
+        view2_page2 = await browser_context.new_page()
+        view2_page3 = await browser_context.new_page()
+
+        await view2_page1.goto(context.make_url('/view-2/'))
+        await view2_page2.goto(context.make_url('/view-2/'))
+        await view2_page3.goto(context.make_url('/view-2/?fire'))
+
+        # broadcast
+        broadcast_page = await browser_context.new_page()
+
+        await broadcast_page.goto(context.make_url('/broadcast/'))
+
+        # wait for all threads to complete
+        await asyncio.sleep(1)
```

Comment:
I think it's better to avoid `sleep`s in tests.
For example I used in my projects following helper https://gist.github.com/maratori/118d1018893b367be8d35ccda996d3af.

Reply:
I am not really happy with this either, but we cant just wait for the call to return, because ``fire_event`` schedules the callbacks to the worker threads anf returnes immediately

Reply:
> but we cant just wait for the call to return

Yes, that's ok.

I meant to use `wait_for` instead of `sleep` + `assert`. Because actually you test that some state will be reached eventually.


One more comment. In my project it was convenient to use `wait_for` above. But here it seems better to create a context manager. Something like
```py
with eventually():
      assert ...
```

Reply:
Ah now i get it. I like that a lot! Could we add this to the lona pytest plugin? I think that would be helpful in test suites for lona applications as well

Reply:
Yeah, why not.

Reply:
Ok. Can you create a PR for it? I would like to leave the sleep in there for now, to get this PR through. You would have a great show case for the new helper.

Reply:
Sure